crates/python/src/py_sync.rs

```diff
@@ -0,0 +1,106 @@
+/// Spawn and block on a future using the pyo3 tokio runtime.
+/// Useful for returning a synchronous `PyResult`.
+///
+///
+/// When used like the following:
+/// ```rs
+/// async fn say_hello(name: String) -> String {
+///     format!("hello {name}")
+/// }
+///
+/// #[pyo3(name="say_hello")]
+/// pub fn py_say_hello(name: String) -> PyResult<String> {
+///     py_sync!(say_hello(name))
+/// }
+/// ```
+///
+/// Becomes the associated "synchronous" python call:
+/// ```py
+/// assert say_hello("Rigetti") == "hello Rigetti"
+/// ```
+macro_rules! py_sync {
+    ($body: expr) => {{
+        let runtime = pyo3_asyncio::tokio::get_runtime();
+        let handle = runtime.spawn($body);
+        runtime
+            .block_on(handle)
+            .map_err(|err| PyRuntimeError::new_err(err.to_string()))?
+    }};
+}
+
+/// Convert a rust future into a Python awaitable using
+/// `pyo3_asyncio::tokio::local_future_into_py`
+macro_rules! py_async {
+    ($py: ident, $body: expr) => {
+        pyo3_asyncio::tokio::local_future_into_py($py, $body)
+    };
+}
+
+/// Given a single implementation of an async function,
+/// create that function as private and two pyfunctions
+/// named after it that can be used to invoke either
+/// blocking or async variants of the same function.
+///
+/// This macro cannot be used when lifetime specifiers are
+/// required, or the pyfunction bodies need additional
+/// parameter handling besides simply calling out to
+/// the underlying `py_async` or `py_sync` macros.
+///
+/// ```rs
+/// // ... becomes python package "things"
+/// create_init_submodule! {
+///     funcs: [
+///         py_do_thing,
+///         py_do_thing_async,
+///     ]
+/// }
+///
+/// py_function_dual! {
+///     #[args(timeout = "None")]
+///     async fn do_thing(timeout: Option<u64>) -> PyResult<String> {
+///         // ... sleep for timeout ...
+///         Ok(String::from("done"))
+///     }
+/// }
+/// ```
+///
+/// becomes in python:
+/// ```py
+/// from things import do_stuff, do_stuff_async
+/// assert do_stuff() == "done"
+/// assert await do_stuff() == "done"
+/// ```
+macro_rules! py_function_dual {
```

Comment:
macros on macros!

Looks good. I'd consider a name that's more obvious in intent - maybe `py_function_sync_async` or something. No doubt we'll run into constraints with this pattern in the future, but (a) writing the functions by hand is an easy escape hatch and (b) it's valuable today and clear in what it does.